docs/src/userguide/language_basics.rst

```diff
@@ -581,44 +900,93 @@ be careful what you do.
 Type Casting
 ------------
 
-Where C uses ``"("`` and ``")"``, Cython uses ``"<"`` and ``">"``. For example::
+The Cython language supports type casting in a similar way as C. Where C uses ``"("`` and ``")"``,
+Cython uses ``"<"`` and ``">"``.  In pure python mode, the ``cython.cast()`` function is used.  For example:
+
+.. tabs::
+
+    .. group-tab:: Pure Python
+
+        .. code-block:: python
+
+            def main():
+                p: cython.p_char
+                q: cython.p_float
+                p = cast(cython.p_char, q)
+
+        When casting a C value to a Python object type or vice versa,
+        Cython will attempt a coercion. Simple examples are casts like ``cast(int, pyobj_value)``,
+        which converts a Python number to a plain C ``int`` value, or ``cast(bytes, charptr_value)``,
+        which copies a C ``char*`` string into a new Python bytes object.
+
+         .. note:: Cython will not prevent a redundant cast, but emits a warning for it.
 
-    cdef char *p
-    cdef float *q
-    p = <char*>q
+        To get the address of some Python object, use a cast to a pointer type
+        like ``cast(p_void, ...)`` or ``cast(pointer(PyObject), ...)``.
+        You can also cast a C pointer back to a Python object reference
+        with ``cast(object, ...)``, or a more specific builtin or extension type
+        (e.g. ``cast(MyExtType, ptr)``). This will increase the reference count of
```

Comment:
```suggestion
        You can also cast a C pointer back to a Python object reference
        with ``cast(object, ...)``, or to a more specific builtin or extension type
        (e.g. ``cast(MyExtType, ptr)``). This will increase the reference count of
```